### ISSUE TITLE
IFC: match formals by name between required function and its implementations

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#17895]
Given the definition of a required function and an implements statement, should the implementation of the required function match the formals' names?

If so, then the following code would be incorrect because `reqFun` is not implemented properly:
```chpl
interface IFC {
  proc reqFun(myFormal: Self);
}

int implements IFC;
proc reqFun(altFormal: int) { }  // 'altFormal' differs from 'myFormal'
```

If we require matching, we need to provide a way for the user to opt out from this requirement. This could be done, for example, by extending the syntax of a "formal" to allow its name be dropped or specified as a single underscore:
```chpl
interface NameAgnostic {
  proc reqFun1(: Self);    // nameless formal
  proc reqFun2(_: Self);   // the formal is '_'
}
```

One case where the user will want to opt out is when the interface and its implementation are defined in different modules and the user cannot modify one or both of them. For example:
```chpl
module Draw {
  interface Drawable {
    proc draw(object: Self);
  }
}

module Shapes {
  record Circle {...}
  proc draw(c: Circle) {...}
}

module Application {
  use Draw, Circle;
  Circle implements Drawable;  // incorrect: draw's formal names differ
}
```

An issue observable in the current implementation is with a "standard ops" module that we would like to claim is implemented in the standard+internal modules for primitive types such as ints:
```chpl
interface StdOps(Val) {
  proc chpl__initCopy(arg: Val, definedConst: bool): Val;
  operator =(ref lhs: Val, rhs: Val): void;
  operator ==(lhs: Val, rhs: Val): bool;
  operator !=(lhs: Val, rhs: Val): bool return !(lhs == rhs);
  proc write(arg: Val): void;
}

int implements StdOps;
```

The implementations of chpl__initCopy in the internal modules use various names for the first formal. This diversity will likely be preserved when we will be able to specify an initializer as an interface-required function instead of chpl__initCopy.

The implementation of write() in the internal modules accepts a tuple, so does not have a proper single formal whose name we can use in `write()` in `StdOps`

**Comment by bradcray** [comment]
I think names should have to match their interfaces by default (for consistency with the rest of the language), but agree there should be a way to opt out of this—ideally using the same approach one would use to opt out of a formal supporting keyword-based argument matching.

**Comment by vasslitvinov** [comment]
Here is the change to enforce formal-name matching:
```diff
--- a/compiler/resolution/interfaceResolution.cpp
+++ b/compiler/resolution/interfaceResolution.cpp
@@ -1318,7 +1318,7 @@
   CallExpr* call = new CallExpr(reqFn->name);
   for_formals(formal, reqFn) {
     ArgSymbol* dup = formal->copy(&fml2act);
-    call->insertAtTail(dup);
+    call->insertAtTail(new NamedExpr(dup->name, new SymExpr(dup)));
     // needed by handleInIntent->checkAnotherFunctionsFormal()
     holder->insertAtHead(new DefExpr(dup));
     formalDups.push_back(dup);
```

I do not want to make this change right now because of the use of `chpl__initCopy()` and `write()` as required functions in some of the existing tests.

**Comment by BryantLam** [comment]
Is this allowed?
```chapel
module Draw {
  interface Drawable {
    proc draw(_: Self); // opt-out
  }
}

module Shapes {
  record Circle {...}
  proc draw(c: Circle) {...}
}

module Application {
  use Draw, Shapes;
  Circle implements Drawable;  // okay

  var circle = new Circle();
  draw(c = circle); // Am I allowed to use an identifier here?

  proc drawIt(x: Drawable) {
    // Cannot use identifier here?
    draw(/*identifier =*/ x);
  }
  drawIt(circle);
}
```

**Comment by BryantLam** [comment]
Also, is there any place where a nameless formal is used today? I really don't like that idea only because it breaks my mental ability to syntactically pattern match.

**Comment by vasslitvinov** [comment]
@BryantLam I agree with your analysis. The implicit rule here is that the binding of actuals to formals is done for each resolution candidate independently of the others.

In the case of `draw(c = circle)` the candidate is `proc draw(c: Circle) {...}` so it is legal to use a named actual.

In the case of `draw(x)`, the only applicable candidate is `proc draw(_: Self);`. I agree it does not make sense to allow named actuals when the formal is not named.

There is also an interesting question of whether `proc draw(_: Self);` is an applicable candidate for the call `draw(circle)` (when the actual is not named). I think it should be applicable, yet less specific than a non-interface-constrained candidate. When one candidate is an interface generic and another a traditional generic, we probably want to choose the interface generic or report ambiguity.

> is there any place where a nameless formal is used today?

No. Neither "opt out" syntax is legal today. Cf. an underscore can be used in a tuple: `var (a,_)  = (3,4);`

>  I really don't like that idea only because it breaks my mental ability to syntactically pattern match.

Somehow I did not follow the connection. In any case, I am open to alternative suggestions for how the user would opt out.

**Comment by bradcray** [comment]
I'll let Vass answer the first set of questions, but to the second:

> Also, is there any place where a nameless formal is used today?

There isn't because a nameless formal on a function definition arguably isn't useful (if you can't refer to it within the function body, why make the formal at all?).   Meanwhile, Chapel doesn't have many other contexts where we have a "function prototype" context, as with interfaces.  The only one I can think of is external procedures.  For example, imagine `extern proc sin(_: real): real;` as a way of saying "C doesn't care what the formal's name is, so why should the `extern` declaration? / sin() callsite?"

However, we have discussed the notion in the language of declaring formals that can't be named at the callsite.  One modest proposal, for example, was:

```chapel
proc foo(_x: int) { ... }  // note formal starts with an underscore
foo(10);                   // OK, doesn't name formal
foo(_x = 10);              // Error: Can't refer to this formal by name at the callsite.
```

My sense of the state of that conversation is:
* "Maybe there should be some way to do this?"
* "Leading underscores are nicely succinct, but seem pretty subtle/convention-based... can we think of something better?"
* ... 

where nobody has (and nobody's cared enough about the feature to push harder on it).  So this interface context might give us an excuse to work on it harder (or decide we don't need it in non-interface contexts).

**Comment by bradcray** [comment]
[oops, my browser didn't show me that Vass beat me to both answers]

> Cf. an underscore can be used in a tuple: var (a,_) = (3,4);

We also use the underscore in module `use` statements to rename a symbol to something that can't be named.  E.g., `use M as _;` prevents `M` from being referred to in the scope.

**Comment by BryantLam** [comment]
> > I really don't like that idea only because it breaks my mental ability to syntactically pattern match.
> 
> Somehow I did not follow the connection.

I'm fine with the underscore variant. My criticism was that the nameless formal "looks weird" enough that my mental compiler warns that _something_ is missing.

I'm good with the other responses. Thanks!

**Comment by mppf** [comment]
> However, we have discussed the notion in the language of declaring formals that can't be named at the callsite. One modest proposal...

We discussed this in issue #11069.

**Comment by bradcray** [comment]
Thanks Michael, I was not finding that yesterday (I was looking for issues that mentioned 'formal').
